src/engine/GroupBy.cpp

```diff
@@ -763,25 +763,35 @@ GroupBy::findAggregates(sparqlExpression::SparqlExpression* expr) {
 
 // _____________________________________________________________________________
 template <class T>
-bool GroupBy::hasType(sparqlExpression::SparqlExpression* expr) {
-  return dynamic_cast<T*>(expr) != nullptr;
+std::optional<T*> GroupBy::hasType(sparqlExpression::SparqlExpression* expr) {
+  auto value = dynamic_cast<T*>(expr);
+  if (value == nullptr)
+    return std::nullopt;
+  else
+    return value;
 }
 
 // _____________________________________________________________________________
 bool GroupBy::isSupportedAggregate(sparqlExpression::SparqlExpression* expr) {
   using namespace sparqlExpression;
 
-  return !(hasType<SumExpression>(expr) || hasType<MinExpression>(expr) ||
-           hasType<MaxExpression>(expr) || hasType<CountExpression>(expr) ||
-           hasType<GroupConcatExpression>(expr));
+  return !(hasType<SumExpression>(expr).has_value() ||
+           hasType<MinExpression>(expr).has_value() ||
+           hasType<MaxExpression>(expr).has_value() ||
+           hasType<CountExpression>(expr).has_value() ||
+           hasType<GroupConcatExpression>(expr).has_value());
 }
 
 // _____________________________________________________________________________
 bool GroupBy::findAggregatesImpl(
     sparqlExpression::SparqlExpression* parent,
     sparqlExpression::SparqlExpression* expr, std::optional<size_t> index,
     std::vector<GroupBy::HashMapAggregateInformation>& info) {
-  if (hasType<sparqlExpression::AvgExpression>(expr)) {
+  if (auto val = hasType<sparqlExpression::AvgExpression>(expr);
+      val.has_value()) {
+    // Do not support distinct aggregates
+    if (val.value()->isDistinct()) return false;
+
```

Comment:
This is fine for now, I suppose this will be refatored anyhow as soon as you support different aggregates.

Reply:
So I shouldn't refactor this into `isUnsupportedAggregate` as described above?

Reply:
I think now that you have them you can immediately refactor them.
Maybe extend your `hasType` to be `hasNonDistinctAggregateType`.
Then `isUnsupportedAggregate` is
```
if (!isAggregate) return false;

return !hasNonDistinctAggregateType<Average>(expr);
```

Which would be the nicest interface and can easily be extended to work with the other expressions.